### PR TITLE
perf(db): cache db handles using OnceLock

### PR DESCRIPTION
**Pull request by DaniPopes**
We only need mutable access once to initialize the value.

Note that this doubles the size of the cache field on non-Linux (`Once` is `AtomicU32` on platforms with `futex` support, otherwise `AtomicPtr`)

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
lgtm

ptal @joshieDo

File: crates/storage/db/src/implementation/mdbx/tx.rs [hunk 1]
```diff
@@ -76,13 +82,22 @@ impl<K: TransactionKind> Tx<K> {
...
+            Some(handle) => Ok(*handle),
+            None => self.open_and_store_db::<T>(slot),
+        }
+    }
+
+    #[cold]
+    fn open_and_store_db<T: Table>(&self, slot: &OnceLock<DBI>) -> Result<DBI, DatabaseError> {
+        match self.inner.open_db(Some(T::NAME)) {
+            Ok(db) => {
+                slot.set(db.dbi()).unwrap();
```
Comment (joshieDo) [inline crates/storage/db/src/implementation/mdbx/tx.rs:97]: can we map_err

**Comment by gakonst** [comment]
What's the perf improvement?

File: crates/storage/db/src/implementation/mdbx/tx.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/storage/db/src/implementation/mdbx/tx.rs:97]: Already tried, this ends up as `map(|db| db.dbi()).map_err(..).inspect(slot.set)`, it's not any better

**Comment by DaniPopes** [comment]
> What's the perf improvement?

OnceLock is cheaper to "lock" than a mutex because we don't need mutable access after the first lazy initialization

**Review by joshieDo — APPROVED** [review]

**Pull request merged**